bigquery-flash-data-sync/internal/pipeline/bqsetup.go

```diff
@@ -0,0 +1,327 @@
+func mysqlTypeToBigQueryType(mysqlType string, logger *zap.Logger) bigquery.FieldType {
+	t := strings.ToUpper(strings.Split(mysqlType, "(")[0])
+	switch t {
+	case "VARCHAR", "CHAR", "TEXT", "TINYTEXT", "MEDIUMTEXT", "LONGTEXT", "ENUM", "SET":
+		return bigquery.StringFieldType
+	case "INT", "TINYINT", "SMALLINT", "MEDIUMINT", "BIGINT", "INTEGER":
+		return bigquery.IntegerFieldType
+	case "FLOAT", "DOUBLE", "DECIMAL", "NUMERIC", "REAL":
+		return bigquery.FloatFieldType
+	case "DATE":
+		return bigquery.DateFieldType
+	case "TIME":
+		return bigquery.TimeFieldType
+	case "DATETIME", "TIMESTAMP":
+		return bigquery.TimestampFieldType
+	case "BOOLEAN", "BOOL", "BIT":
+		return bigquery.BooleanFieldType
+	case "BLOB", "TINYBLOB", "MEDIUMBLOB", "LONGBLOB", "BINARY", "VARBINARY":
+		return bigquery.BytesFieldType
+	case "JSON":
+		return bigquery.JSONFieldType
+	default:
+		logger.Warn("Unknown MySQL type, defaulting to STRING",
+			zap.String("mysql_type", mysqlType),
+			zap.String("default_type", "STRING"))
+		return bigquery.StringFieldType
+	}
+}
+
+// postgresTypeToBigQueryType maps common PostgreSQL database types to BigQuery types.
+func postgresTypeToBigQueryType(pgType string, logger *zap.Logger) bigquery.FieldType {
+	t := strings.ToUpper(strings.Split(pgType, "(")[0])
+	t = strings.TrimSuffix(t, "[]")
+
+	switch t {
+	case "VARCHAR", "CHAR", "CHARACTER", "CHARACTER VARYING", "TEXT", "NAME", "UUID", "CITEXT":
+		return bigquery.StringFieldType
+	case "INT", "INT2", "INT4", "INT8", "INTEGER", "SMALLINT", "BIGINT", "SERIAL", "BIGSERIAL", "SMALLSERIAL":
+		return bigquery.IntegerFieldType
+	case "FLOAT", "FLOAT4", "FLOAT8", "DOUBLE", "DOUBLE PRECISION", "DECIMAL", "NUMERIC", "REAL", "MONEY":
+		return bigquery.FloatFieldType
+	case "DATE":
+		return bigquery.DateFieldType
+	case "TIME", "TIMETZ", "TIME WITH TIME ZONE", "TIME WITHOUT TIME ZONE":
+		return bigquery.TimeFieldType
+	case "TIMESTAMP", "TIMESTAMPTZ", "TIMESTAMP WITH TIME ZONE", "TIMESTAMP WITHOUT TIME ZONE":
+		return bigquery.TimestampFieldType
+	case "BOOLEAN", "BOOL":
+		return bigquery.BooleanFieldType
+	case "BYTEA":
+		return bigquery.BytesFieldType
+	case "JSON", "JSONB":
+		return bigquery.JSONFieldType
+	case "INET", "CIDR", "MACADDR", "MACADDR8":
+		return bigquery.StringFieldType
+	case "INTERVAL":
+		return bigquery.StringFieldType
+	case "POINT", "LINE", "LSEG", "BOX", "PATH", "POLYGON", "CIRCLE":
+		return bigquery.StringFieldType
+	default:
+		logger.Warn("Unknown PostgreSQL type, defaulting to STRING",
+			zap.String("postgres_type", pgType),
+			zap.String("default_type", "STRING"))
+		return bigquery.StringFieldType
+	}
+}
+
+// InferSchemaFromMySQL connects to the source DB, runs a LIMIT 1 query,
+// and builds a BigQuery Schema based on the returned column types.
+func InferSchemaFromMySQL(db *sql.DB, query string, logger *zap.Logger) (bigquery.Schema, error) {
+	logger.Debug("Inferring schema from MySQL database")
```

Comment:
Let's make the debugs more specific project wide so we can hopefully find the location of a bug
```suggestion
	logger.Debug("Inferring schema from MySQL database", zap.String("database", db.Name)) // Or related
```

Reply:
Ack